hepcrawl/spiders/elsevier.py

```diff
@@ -0,0 +1,235 @@
+import glob
+import os
+import shutil
+import tempfile
+import xml.etree.ElementTree as et
+import zipfile
+from io import BytesIO
+
+import boto3
+import scrapy
+from scrapy import Request, Spider
+
+from . import StatefulSpider
+from ..parsers import ElsevierParser
+from ..utils import ParsedItem, strict_kwargs
+
+
+class ElsevierSpider(StatefulSpider):
+    name = "elsevier"
+    start_urls = []
+
+    @strict_kwargs
+    def __init__(
+        self,
+        acces_key_id,
+        secret_access_key,
+        packages_bucket_name,
+        files_bucket_name,
+        elsevier_consyn_url,
+        s3_host="https://s3.cern.ch",
+        *args,
+        **kwargs
+    ):
+        super(ElsevierSpider, self).__init__(*args, **kwargs)
+        self.access_key_id = acces_key_id
+        self.secret_access_key = secret_access_key
+        self.packages_bucket_name = packages_bucket_name
+        self.files_bucket_name = files_bucket_name
+        self.elsevier_consyn_url = elsevier_consyn_url
+        self.new_packages = set()
+        self.new_xml_files = set()
+        self.s3_host = s3_host
+
+        if not (
+            self.access_key_id,
+            self.secret_access_key,
+            self.packages_bucket_name,
+            self.files_bucket_name,
+        ):
+            raise Exception("Missing parametrs necessary to establish s3 connection")
+        else:
+            self.s3_connection = self.create_s3_connection()
+            self.s3_packages_bucket_conn = self.s3_bucket_connection(
+                self.packages_bucket_name
+            )
+            self.s3_files_bucket_conn = self.s3_bucket_connection(
+                self.files_bucket_name
+            )
+            self.s3_client = self.connect_s3_client()
+
+    def create_s3_connection(self):
+        session = boto3.Session(
+            aws_access_key_id=self.access_key_id,
+            aws_secret_access_key=self.secret_access_key,
+        )
+        s3 = session.resource("s3", endpoint_url=self.s3_host)
+        return s3
+
+    def s3_bucket_connection(self, bucket_name):
+        bucket_connection = self.s3_connection.Bucket(bucket_name)
+        return bucket_connection
+
+    def connect_s3_client(self):
+        s3_client = boto3.client(
+            "s3",
+            aws_access_key_id=self.access_key_id,
+            aws_secret_access_key=self.secret_access_key,
+            endpoint_url=self.s3_host,
+        )
+        return s3_client
+
+    def create_presigned_url(self, bucket, file, method):
+        url = self.s3_client.generate_presigned_url(
+            ClientMethod=method,
+            Params={"Bucket": bucket, "Key": file},
+            ExpiresIn=920000,
+        )
+        return url
+
+    def _get_keys_names_from_bucket(self):
+        keys = set([key.key for key in self.s3_packages_bucket_conn.objects.all()])
```

Comment:
this will become very slow as time goes on. Maybe it's better to check whether each file exists by doing a request for each zip file in the feed?